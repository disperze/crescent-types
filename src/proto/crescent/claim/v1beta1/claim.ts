/* eslint-disable */
import Long from "long";
import * as _m0 from "protobufjs/minimal";
import { Timestamp } from "../../../google/protobuf/timestamp";
import { Coin } from "../../../cosmos/base/v1beta1/coin";

export const protobufPackage = "crescent.claim.v1beta1";

/** ConditionType defines the type of condition that a recipient must execute in order to receive a claimable amount. */
export enum ConditionType {
  /** CONDITION_TYPE_UNSPECIFIED - CONDITION_TYPE_UNSPECIFIED specifies an unknown condition type */
  CONDITION_TYPE_UNSPECIFIED = 0,
  /** CONDITION_TYPE_DEPOSIT - CONDITION_TYPE_DEPOSIT specifies deposit condition type */
  CONDITION_TYPE_DEPOSIT = 1,
  /** CONDITION_TYPE_SWAP - CONDITION_TYPE_SWAP specifies swap condition type */
  CONDITION_TYPE_SWAP = 2,
  /** CONDITION_TYPE_LIQUIDSTAKE - CONDITION_TYPE_LIQUIDSTAKE specifies liquid stake condition type */
  CONDITION_TYPE_LIQUIDSTAKE = 3,
  /** CONDITION_TYPE_VOTE - CONDITION_TYPE_VOTE specifies governance vote condition type */
  CONDITION_TYPE_VOTE = 4,
  UNRECOGNIZED = -1,
}

export function conditionTypeFromJSON(object: any): ConditionType {
  switch (object) {
    case 0:
    case "CONDITION_TYPE_UNSPECIFIED":
      return ConditionType.CONDITION_TYPE_UNSPECIFIED;
    case 1:
    case "CONDITION_TYPE_DEPOSIT":
      return ConditionType.CONDITION_TYPE_DEPOSIT;
    case 2:
    case "CONDITION_TYPE_SWAP":
      return ConditionType.CONDITION_TYPE_SWAP;
    case 3:
    case "CONDITION_TYPE_LIQUIDSTAKE":
      return ConditionType.CONDITION_TYPE_LIQUIDSTAKE;
    case 4:
    case "CONDITION_TYPE_VOTE":
      return ConditionType.CONDITION_TYPE_VOTE;
    case -1:
    case "UNRECOGNIZED":
    default:
      return ConditionType.UNRECOGNIZED;
  }
}

export function conditionTypeToJSON(object: ConditionType): string {
  switch (object) {
    case ConditionType.CONDITION_TYPE_UNSPECIFIED:
      return "CONDITION_TYPE_UNSPECIFIED";
    case ConditionType.CONDITION_TYPE_DEPOSIT:
      return "CONDITION_TYPE_DEPOSIT";
    case ConditionType.CONDITION_TYPE_SWAP:
      return "CONDITION_TYPE_SWAP";
    case ConditionType.CONDITION_TYPE_LIQUIDSTAKE:
      return "CONDITION_TYPE_LIQUIDSTAKE";
    case ConditionType.CONDITION_TYPE_VOTE:
      return "CONDITION_TYPE_VOTE";
    default:
      return "UNKNOWN";
  }
}

/** Airdrop defines airdrop information. */
export interface Airdrop {
  /** id specifies index of the airdrop */
  id: Long;
  /**
   * source_address defines the bech32-encoded source address
   * where the source of coins from
   */
  sourceAddress: string;
  /** conditions specifies a list of conditions */
  conditions: ConditionType[];
  /** start_time specifies the start time of the airdrop */
  startTime?: Date;
  /** end_time specifies the start time of the airdrop */
  endTime?: Date;
}

/** ClaimRecord defines claim record that corresponds to the airdrop. */
export interface ClaimRecord {
  /** airdrop_id specifies airdrop id */
  airdropId: Long;
  /** recipient specifies the bech32-encoded address that is eligible to claim airdrop */
  recipient: string;
  /** initial_claimable_coins specifies the initial claimable coins */
  initialClaimableCoins: Coin[];
  /** claimable_coins specifies the unclaimed claimable coins */
  claimableCoins: Coin[];
  /**
   * claimed_conditions specifies a list of condition types
   * initial values are empty and each condition type gets appended when claim is successfully executed
   */
  claimedConditions: ConditionType[];
}

function createBaseAirdrop(): Airdrop {
  return {
    id: Long.UZERO,
    sourceAddress: "",
    conditions: [],
    startTime: undefined,
    endTime: undefined,
  };
}

export const Airdrop = {
  encode(
    message: Airdrop,
    writer: _m0.Writer = _m0.Writer.create()
  ): _m0.Writer {
    if (!message.id.isZero()) {
      writer.uint32(8).uint64(message.id);
    }
    if (message.sourceAddress !== "") {
      writer.uint32(18).string(message.sourceAddress);
    }
    writer.uint32(26).fork();
    for (const v of message.conditions) {
      writer.int32(v);
    }
    writer.ldelim();
    if (message.startTime !== undefined) {
      Timestamp.encode(
        toTimestamp(message.startTime),
        writer.uint32(34).fork()
      ).ldelim();
    }
    if (message.endTime !== undefined) {
      Timestamp.encode(
        toTimestamp(message.endTime),
        writer.uint32(42).fork()
      ).ldelim();
    }
    return writer;
  },

  decode(input: _m0.Reader | Uint8Array, length?: number): Airdrop {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseAirdrop();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          message.id = reader.uint64() as Long;
          break;
        case 2:
          message.sourceAddress = reader.string();
          break;
        case 3:
          if ((tag & 7) === 2) {
            const end2 = reader.uint32() + reader.pos;
            while (reader.pos < end2) {
              message.conditions.push(reader.int32() as any);
            }
          } else {
            message.conditions.push(reader.int32() as any);
          }
          break;
        case 4:
          message.startTime = fromTimestamp(
            Timestamp.decode(reader, reader.uint32())
          );
          break;
        case 5:
          message.endTime = fromTimestamp(
            Timestamp.decode(reader, reader.uint32())
          );
          break;
        default:
          reader.skipType(tag & 7);
          break;
      }
    }
    return message;
  },

  fromJSON(object: any): Airdrop {
    return {
      id: isSet(object.id) ? Long.fromString(object.id) : Long.UZERO,
      sourceAddress: isSet(object.sourceAddress)
        ? String(object.sourceAddress)
        : "",
      conditions: Array.isArray(object?.conditions)
        ? object.conditions.map((e: any) => conditionTypeFromJSON(e))
        : [],
      startTime: isSet(object.startTime)
        ? fromJsonTimestamp(object.startTime)
        : undefined,
      endTime: isSet(object.endTime)
        ? fromJsonTimestamp(object.endTime)
        : undefined,
    };
  },

  toJSON(message: Airdrop): unknown {
    const obj: any = {};
    message.id !== undefined &&
      (obj.id = (message.id || Long.UZERO).toString());
    message.sourceAddress !== undefined &&
      (obj.sourceAddress = message.sourceAddress);
    if (message.conditions) {
      obj.conditions = message.conditions.map((e) => conditionTypeToJSON(e));
    } else {
      obj.conditions = [];
    }
    message.startTime !== undefined &&
      (obj.startTime = message.startTime.toISOString());
    message.endTime !== undefined &&
      (obj.endTime = message.endTime.toISOString());
    return obj;
  },

  fromPartial<I extends Exact<DeepPartial<Airdrop>, I>>(object: I): Airdrop {
    const message = createBaseAirdrop();
    message.id =
      object.id !== undefined && object.id !== null
        ? Long.fromValue(object.id)
        : Long.UZERO;
    message.sourceAddress = object.sourceAddress ?? "";
    message.conditions = object.conditions?.map((e) => e) || [];
    message.startTime = object.startTime ?? undefined;
    message.endTime = object.endTime ?? undefined;
    return message;
  },
};

function createBaseClaimRecord(): ClaimRecord {
  return {
    airdropId: Long.UZERO,
    recipient: "",
    initialClaimableCoins: [],
    claimableCoins: [],
    claimedConditions: [],
  };
}

export const ClaimRecord = {
  encode(
    message: ClaimRecord,
    writer: _m0.Writer = _m0.Writer.create()
  ): _m0.Writer {
    if (!message.airdropId.isZero()) {
      writer.uint32(8).uint64(message.airdropId);
    }
    if (message.recipient !== "") {
      writer.uint32(18).string(message.recipient);
    }
    for (const v of message.initialClaimableCoins) {
      Coin.encode(v!, writer.uint32(26).fork()).ldelim();
    }
    for (const v of message.claimableCoins) {
      Coin.encode(v!, writer.uint32(34).fork()).ldelim();
    }
    writer.uint32(42).fork();
    for (const v of message.claimedConditions) {
      writer.int32(v);
    }
    writer.ldelim();
    return writer;
  },

  decode(input: _m0.Reader | Uint8Array, length?: number): ClaimRecord {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseClaimRecord();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          message.airdropId = reader.uint64() as Long;
          break;
        case 2:
          message.recipient = reader.string();
          break;
        case 3:
          message.initialClaimableCoins.push(
            Coin.decode(reader, reader.uint32())
          );
          break;
        case 4:
          message.claimableCoins.push(Coin.decode(reader, reader.uint32()));
          break;
        case 5:
          if ((tag & 7) === 2) {
            const end2 = reader.uint32() + reader.pos;
            while (reader.pos < end2) {
              message.claimedConditions.push(reader.int32() as any);
            }
          } else {
            message.claimedConditions.push(reader.int32() as any);
          }
          break;
        default:
          reader.skipType(tag & 7);
          break;
      }
    }
    return message;
  },

  fromJSON(object: any): ClaimRecord {
    return {
      airdropId: isSet(object.airdropId)
        ? Long.fromString(object.airdropId)
        : Long.UZERO,
      recipient: isSet(object.recipient) ? String(object.recipient) : "",
      initialClaimableCoins: Array.isArray(object?.initialClaimableCoins)
        ? object.initialClaimableCoins.map((e: any) => Coin.fromJSON(e))
        : [],
      claimableCoins: Array.isArray(object?.claimableCoins)
        ? object.claimableCoins.map((e: any) => Coin.fromJSON(e))
        : [],
      claimedConditions: Array.isArray(object?.claimedConditions)
        ? object.claimedConditions.map((e: any) => conditionTypeFromJSON(e))
        : [],
    };
  },

  toJSON(message: ClaimRecord): unknown {
    const obj: any = {};
    message.airdropId !== undefined &&
      (obj.airdropId = (message.airdropId || Long.UZERO).toString());
    message.recipient !== undefined && (obj.recipient = message.recipient);
    if (message.initialClaimableCoins) {
      obj.initialClaimableCoins = message.initialClaimableCoins.map((e) =>
        e ? Coin.toJSON(e) : undefined
      );
    } else {
      obj.initialClaimableCoins = [];
    }
    if (message.claimableCoins) {
      obj.claimableCoins = message.claimableCoins.map((e) =>
        e ? Coin.toJSON(e) : undefined
      );
    } else {
      obj.claimableCoins = [];
    }
    if (message.claimedConditions) {
      obj.claimedConditions = message.claimedConditions.map((e) =>
        conditionTypeToJSON(e)
      );
    } else {
      obj.claimedConditions = [];
    }
    return obj;
  },

  fromPartial<I extends Exact<DeepPartial<ClaimRecord>, I>>(
    object: I
  ): ClaimRecord {
    const message = createBaseClaimRecord();
    message.airdropId =
      object.airdropId !== undefined && object.airdropId !== null
        ? Long.fromValue(object.airdropId)
        : Long.UZERO;
    message.recipient = object.recipient ?? "";
    message.initialClaimableCoins =
      object.initialClaimableCoins?.map((e) => Coin.fromPartial(e)) || [];
    message.claimableCoins =
      object.claimableCoins?.map((e) => Coin.fromPartial(e)) || [];
    message.claimedConditions = object.claimedConditions?.map((e) => e) || [];
    return message;
  },
};

type Builtin =
  | Date
  | Function
  | Uint8Array
  | string
  | number
  | boolean
  | undefined;

export type DeepPartial<T> = T extends Builtin
  ? T
  : T extends Long
  ? string | number | Long
  : T extends Array<infer U>
  ? Array<DeepPartial<U>>
  : T extends ReadonlyArray<infer U>
  ? ReadonlyArray<DeepPartial<U>>
  : T extends {}
  ? { [K in keyof T]?: DeepPartial<T[K]> }
  : Partial<T>;

type KeysOfUnion<T> = T extends T ? keyof T : never;
export type Exact<P, I extends P> = P extends Builtin
  ? P
  : P & { [K in keyof P]: Exact<P[K], I[K]> } & Record<
        Exclude<keyof I, KeysOfUnion<P>>,
        never
      >;

function toTimestamp(date: Date): Timestamp {
  const seconds = numberToLong(date.getTime() / 1_000);
  const nanos = (date.getTime() % 1_000) * 1_000_000;
  return { seconds, nanos };
}

function fromTimestamp(t: Timestamp): Date {
  let millis = t.seconds.toNumber() * 1_000;
  millis += t.nanos / 1_000_000;
  return new Date(millis);
}

function fromJsonTimestamp(o: any): Date {
  if (o instanceof Date) {
    return o;
  } else if (typeof o === "string") {
    return new Date(o);
  } else {
    return fromTimestamp(Timestamp.fromJSON(o));
  }
}

function numberToLong(number: number) {
  return Long.fromNumber(number);
}

if (_m0.util.Long !== Long) {
  _m0.util.Long = Long as any;
  _m0.configure();
}

function isSet(value: any): boolean {
  return value !== null && value !== undefined;
}
