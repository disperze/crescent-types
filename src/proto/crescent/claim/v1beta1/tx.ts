/* eslint-disable */
import Long from "long";
import * as _m0 from "protobufjs/minimal";
import {
  ConditionType,
  conditionTypeFromJSON,
  conditionTypeToJSON,
} from "../../../crescent/claim/v1beta1/claim";

export const protobufPackage = "crescent.claim.v1beta1";

/** MsgClaim defines a SDK message for claiming claimable amount. */
export interface MsgClaim {
  /** airdrop_id specifies index of the airdrop */
  airdropId: Long;
  /** recipient specifies the bech32-encoded address that is eligible to claim airdrop */
  recipient: string;
  /** condition_type specifies the condition type */
  conditionType: ConditionType;
}

export interface MsgClaimResponse {}

function createBaseMsgClaim(): MsgClaim {
  return { airdropId: Long.UZERO, recipient: "", conditionType: 0 };
}

export const MsgClaim = {
  encode(
    message: MsgClaim,
    writer: _m0.Writer = _m0.Writer.create()
  ): _m0.Writer {
    if (!message.airdropId.isZero()) {
      writer.uint32(8).uint64(message.airdropId);
    }
    if (message.recipient !== "") {
      writer.uint32(18).string(message.recipient);
    }
    if (message.conditionType !== 0) {
      writer.uint32(24).int32(message.conditionType);
    }
    return writer;
  },

  decode(input: _m0.Reader | Uint8Array, length?: number): MsgClaim {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseMsgClaim();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          message.airdropId = reader.uint64() as Long;
          break;
        case 2:
          message.recipient = reader.string();
          break;
        case 3:
          message.conditionType = reader.int32() as any;
          break;
        default:
          reader.skipType(tag & 7);
          break;
      }
    }
    return message;
  },

  fromJSON(object: any): MsgClaim {
    return {
      airdropId: isSet(object.airdropId)
        ? Long.fromString(object.airdropId)
        : Long.UZERO,
      recipient: isSet(object.recipient) ? String(object.recipient) : "",
      conditionType: isSet(object.conditionType)
        ? conditionTypeFromJSON(object.conditionType)
        : 0,
    };
  },

  toJSON(message: MsgClaim): unknown {
    const obj: any = {};
    message.airdropId !== undefined &&
      (obj.airdropId = (message.airdropId || Long.UZERO).toString());
    message.recipient !== undefined && (obj.recipient = message.recipient);
    message.conditionType !== undefined &&
      (obj.conditionType = conditionTypeToJSON(message.conditionType));
    return obj;
  },

  fromPartial<I extends Exact<DeepPartial<MsgClaim>, I>>(object: I): MsgClaim {
    const message = createBaseMsgClaim();
    message.airdropId =
      object.airdropId !== undefined && object.airdropId !== null
        ? Long.fromValue(object.airdropId)
        : Long.UZERO;
    message.recipient = object.recipient ?? "";
    message.conditionType = object.conditionType ?? 0;
    return message;
  },
};

function createBaseMsgClaimResponse(): MsgClaimResponse {
  return {};
}

export const MsgClaimResponse = {
  encode(
    _: MsgClaimResponse,
    writer: _m0.Writer = _m0.Writer.create()
  ): _m0.Writer {
    return writer;
  },

  decode(input: _m0.Reader | Uint8Array, length?: number): MsgClaimResponse {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseMsgClaimResponse();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        default:
          reader.skipType(tag & 7);
          break;
      }
    }
    return message;
  },

  fromJSON(_: any): MsgClaimResponse {
    return {};
  },

  toJSON(_: MsgClaimResponse): unknown {
    const obj: any = {};
    return obj;
  },

  fromPartial<I extends Exact<DeepPartial<MsgClaimResponse>, I>>(
    _: I
  ): MsgClaimResponse {
    const message = createBaseMsgClaimResponse();
    return message;
  },
};

/** Msg defines the Msg service. */
export interface Msg {
  Claim(request: MsgClaim): Promise<MsgClaimResponse>;
}

export class MsgClientImpl implements Msg {
  private readonly rpc: Rpc;
  constructor(rpc: Rpc) {
    this.rpc = rpc;
    this.Claim = this.Claim.bind(this);
  }
  Claim(request: MsgClaim): Promise<MsgClaimResponse> {
    const data = MsgClaim.encode(request).finish();
    const promise = this.rpc.request(
      "crescent.claim.v1beta1.Msg",
      "Claim",
      data
    );
    return promise.then((data) =>
      MsgClaimResponse.decode(new _m0.Reader(data))
    );
  }
}

interface Rpc {
  request(
    service: string,
    method: string,
    data: Uint8Array
  ): Promise<Uint8Array>;
}

type Builtin =
  | Date
  | Function
  | Uint8Array
  | string
  | number
  | boolean
  | undefined;

export type DeepPartial<T> = T extends Builtin
  ? T
  : T extends Long
  ? string | number | Long
  : T extends Array<infer U>
  ? Array<DeepPartial<U>>
  : T extends ReadonlyArray<infer U>
  ? ReadonlyArray<DeepPartial<U>>
  : T extends {}
  ? { [K in keyof T]?: DeepPartial<T[K]> }
  : Partial<T>;

type KeysOfUnion<T> = T extends T ? keyof T : never;
export type Exact<P, I extends P> = P extends Builtin
  ? P
  : P & { [K in keyof P]: Exact<P[K], I[K]> } & Record<
        Exclude<keyof I, KeysOfUnion<P>>,
        never
      >;

if (_m0.util.Long !== Long) {
  _m0.util.Long = Long as any;
  _m0.configure();
}

function isSet(value: any): boolean {
  return value !== null && value !== undefined;
}
