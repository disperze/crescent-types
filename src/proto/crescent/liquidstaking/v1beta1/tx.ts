/* eslint-disable */
import Long from "long";
import * as _m0 from "protobufjs/minimal";
import { Coin } from "../../../cosmos/base/v1beta1/coin";
import { Timestamp } from "../../../google/protobuf/timestamp";

export const protobufPackage = "crescent.liquidstaking.v1beta1";

/**
 * MsgLiquidStake defines a SDK message for performing a liquid stake of coins
 * from a delegator to whitelisted validators.
 */
export interface MsgLiquidStake {
  delegatorAddress: string;
  amount?: Coin;
}

/** MsgLiquidStakeResponse defines the Msg/LiquidStake response type. */
export interface MsgLiquidStakeResponse {}

/**
 * MsgLiquidUnstake defines a SDK message for performing an undelegation of liquid staking from a
 * delegate.
 */
export interface MsgLiquidUnstake {
  delegatorAddress: string;
  amount?: Coin;
}

/** MsgLiquidUnstakeResponse defines the Msg/LiquidUnstake response type. */
export interface MsgLiquidUnstakeResponse {
  completionTime?: Date;
}

function createBaseMsgLiquidStake(): MsgLiquidStake {
  return { delegatorAddress: "", amount: undefined };
}

export const MsgLiquidStake = {
  encode(
    message: MsgLiquidStake,
    writer: _m0.Writer = _m0.Writer.create()
  ): _m0.Writer {
    if (message.delegatorAddress !== "") {
      writer.uint32(10).string(message.delegatorAddress);
    }
    if (message.amount !== undefined) {
      Coin.encode(message.amount, writer.uint32(18).fork()).ldelim();
    }
    return writer;
  },

  decode(input: _m0.Reader | Uint8Array, length?: number): MsgLiquidStake {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseMsgLiquidStake();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          message.delegatorAddress = reader.string();
          break;
        case 2:
          message.amount = Coin.decode(reader, reader.uint32());
          break;
        default:
          reader.skipType(tag & 7);
          break;
      }
    }
    return message;
  },

  fromJSON(object: any): MsgLiquidStake {
    return {
      delegatorAddress: isSet(object.delegatorAddress)
        ? String(object.delegatorAddress)
        : "",
      amount: isSet(object.amount) ? Coin.fromJSON(object.amount) : undefined,
    };
  },

  toJSON(message: MsgLiquidStake): unknown {
    const obj: any = {};
    message.delegatorAddress !== undefined &&
      (obj.delegatorAddress = message.delegatorAddress);
    message.amount !== undefined &&
      (obj.amount = message.amount ? Coin.toJSON(message.amount) : undefined);
    return obj;
  },

  fromPartial<I extends Exact<DeepPartial<MsgLiquidStake>, I>>(
    object: I
  ): MsgLiquidStake {
    const message = createBaseMsgLiquidStake();
    message.delegatorAddress = object.delegatorAddress ?? "";
    message.amount =
      object.amount !== undefined && object.amount !== null
        ? Coin.fromPartial(object.amount)
        : undefined;
    return message;
  },
};

function createBaseMsgLiquidStakeResponse(): MsgLiquidStakeResponse {
  return {};
}

export const MsgLiquidStakeResponse = {
  encode(
    _: MsgLiquidStakeResponse,
    writer: _m0.Writer = _m0.Writer.create()
  ): _m0.Writer {
    return writer;
  },

  decode(
    input: _m0.Reader | Uint8Array,
    length?: number
  ): MsgLiquidStakeResponse {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseMsgLiquidStakeResponse();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        default:
          reader.skipType(tag & 7);
          break;
      }
    }
    return message;
  },

  fromJSON(_: any): MsgLiquidStakeResponse {
    return {};
  },

  toJSON(_: MsgLiquidStakeResponse): unknown {
    const obj: any = {};
    return obj;
  },

  fromPartial<I extends Exact<DeepPartial<MsgLiquidStakeResponse>, I>>(
    _: I
  ): MsgLiquidStakeResponse {
    const message = createBaseMsgLiquidStakeResponse();
    return message;
  },
};

function createBaseMsgLiquidUnstake(): MsgLiquidUnstake {
  return { delegatorAddress: "", amount: undefined };
}

export const MsgLiquidUnstake = {
  encode(
    message: MsgLiquidUnstake,
    writer: _m0.Writer = _m0.Writer.create()
  ): _m0.Writer {
    if (message.delegatorAddress !== "") {
      writer.uint32(10).string(message.delegatorAddress);
    }
    if (message.amount !== undefined) {
      Coin.encode(message.amount, writer.uint32(18).fork()).ldelim();
    }
    return writer;
  },

  decode(input: _m0.Reader | Uint8Array, length?: number): MsgLiquidUnstake {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseMsgLiquidUnstake();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          message.delegatorAddress = reader.string();
          break;
        case 2:
          message.amount = Coin.decode(reader, reader.uint32());
          break;
        default:
          reader.skipType(tag & 7);
          break;
      }
    }
    return message;
  },

  fromJSON(object: any): MsgLiquidUnstake {
    return {
      delegatorAddress: isSet(object.delegatorAddress)
        ? String(object.delegatorAddress)
        : "",
      amount: isSet(object.amount) ? Coin.fromJSON(object.amount) : undefined,
    };
  },

  toJSON(message: MsgLiquidUnstake): unknown {
    const obj: any = {};
    message.delegatorAddress !== undefined &&
      (obj.delegatorAddress = message.delegatorAddress);
    message.amount !== undefined &&
      (obj.amount = message.amount ? Coin.toJSON(message.amount) : undefined);
    return obj;
  },

  fromPartial<I extends Exact<DeepPartial<MsgLiquidUnstake>, I>>(
    object: I
  ): MsgLiquidUnstake {
    const message = createBaseMsgLiquidUnstake();
    message.delegatorAddress = object.delegatorAddress ?? "";
    message.amount =
      object.amount !== undefined && object.amount !== null
        ? Coin.fromPartial(object.amount)
        : undefined;
    return message;
  },
};

function createBaseMsgLiquidUnstakeResponse(): MsgLiquidUnstakeResponse {
  return { completionTime: undefined };
}

export const MsgLiquidUnstakeResponse = {
  encode(
    message: MsgLiquidUnstakeResponse,
    writer: _m0.Writer = _m0.Writer.create()
  ): _m0.Writer {
    if (message.completionTime !== undefined) {
      Timestamp.encode(
        toTimestamp(message.completionTime),
        writer.uint32(10).fork()
      ).ldelim();
    }
    return writer;
  },

  decode(
    input: _m0.Reader | Uint8Array,
    length?: number
  ): MsgLiquidUnstakeResponse {
    const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
    let end = length === undefined ? reader.len : reader.pos + length;
    const message = createBaseMsgLiquidUnstakeResponse();
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          message.completionTime = fromTimestamp(
            Timestamp.decode(reader, reader.uint32())
          );
          break;
        default:
          reader.skipType(tag & 7);
          break;
      }
    }
    return message;
  },

  fromJSON(object: any): MsgLiquidUnstakeResponse {
    return {
      completionTime: isSet(object.completionTime)
        ? fromJsonTimestamp(object.completionTime)
        : undefined,
    };
  },

  toJSON(message: MsgLiquidUnstakeResponse): unknown {
    const obj: any = {};
    message.completionTime !== undefined &&
      (obj.completionTime = message.completionTime.toISOString());
    return obj;
  },

  fromPartial<I extends Exact<DeepPartial<MsgLiquidUnstakeResponse>, I>>(
    object: I
  ): MsgLiquidUnstakeResponse {
    const message = createBaseMsgLiquidUnstakeResponse();
    message.completionTime = object.completionTime ?? undefined;
    return message;
  },
};

/** Msg defines the liquid staking Msg service. */
export interface Msg {
  /**
   * LiquidStake defines a method for performing a delegation of coins
   * from a delegator to whitelisted validators.
   */
  LiquidStake(request: MsgLiquidStake): Promise<MsgLiquidStakeResponse>;
  /**
   * LiquidUnstake defines a method for performing an undelegation of liquid staking from a
   * delegate.
   */
  LiquidUnstake(request: MsgLiquidUnstake): Promise<MsgLiquidUnstakeResponse>;
}

export class MsgClientImpl implements Msg {
  private readonly rpc: Rpc;
  constructor(rpc: Rpc) {
    this.rpc = rpc;
    this.LiquidStake = this.LiquidStake.bind(this);
    this.LiquidUnstake = this.LiquidUnstake.bind(this);
  }
  LiquidStake(request: MsgLiquidStake): Promise<MsgLiquidStakeResponse> {
    const data = MsgLiquidStake.encode(request).finish();
    const promise = this.rpc.request(
      "crescent.liquidstaking.v1beta1.Msg",
      "LiquidStake",
      data
    );
    return promise.then((data) =>
      MsgLiquidStakeResponse.decode(new _m0.Reader(data))
    );
  }

  LiquidUnstake(request: MsgLiquidUnstake): Promise<MsgLiquidUnstakeResponse> {
    const data = MsgLiquidUnstake.encode(request).finish();
    const promise = this.rpc.request(
      "crescent.liquidstaking.v1beta1.Msg",
      "LiquidUnstake",
      data
    );
    return promise.then((data) =>
      MsgLiquidUnstakeResponse.decode(new _m0.Reader(data))
    );
  }
}

interface Rpc {
  request(
    service: string,
    method: string,
    data: Uint8Array
  ): Promise<Uint8Array>;
}

type Builtin =
  | Date
  | Function
  | Uint8Array
  | string
  | number
  | boolean
  | undefined;

export type DeepPartial<T> = T extends Builtin
  ? T
  : T extends Long
  ? string | number | Long
  : T extends Array<infer U>
  ? Array<DeepPartial<U>>
  : T extends ReadonlyArray<infer U>
  ? ReadonlyArray<DeepPartial<U>>
  : T extends {}
  ? { [K in keyof T]?: DeepPartial<T[K]> }
  : Partial<T>;

type KeysOfUnion<T> = T extends T ? keyof T : never;
export type Exact<P, I extends P> = P extends Builtin
  ? P
  : P & { [K in keyof P]: Exact<P[K], I[K]> } & Record<
        Exclude<keyof I, KeysOfUnion<P>>,
        never
      >;

function toTimestamp(date: Date): Timestamp {
  const seconds = numberToLong(date.getTime() / 1_000);
  const nanos = (date.getTime() % 1_000) * 1_000_000;
  return { seconds, nanos };
}

function fromTimestamp(t: Timestamp): Date {
  let millis = t.seconds.toNumber() * 1_000;
  millis += t.nanos / 1_000_000;
  return new Date(millis);
}

function fromJsonTimestamp(o: any): Date {
  if (o instanceof Date) {
    return o;
  } else if (typeof o === "string") {
    return new Date(o);
  } else {
    return fromTimestamp(Timestamp.fromJSON(o));
  }
}

function numberToLong(number: number) {
  return Long.fromNumber(number);
}

if (_m0.util.Long !== Long) {
  _m0.util.Long = Long as any;
  _m0.configure();
}

function isSet(value: any): boolean {
  return value !== null && value !== undefined;
}
