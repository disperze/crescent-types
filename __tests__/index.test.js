import { MsgClaim } from '../src/proto/crescent/claim/v1beta1/tx';
import { ConditionType } from '../src/proto/crescent/claim/v1beta1/claim';
import Long from 'long';

it('fromPartial', async () => {
  expect(
    MsgClaim.fromPartial({
      airdropId: Long.fromInt(1),
      recipient: "cosmos1y54exmx84cqtasvjnskf9fpx9c0rk9tk8fv4",
      conditionType: ConditionType.CONDITION_TYPE_VOTE
    })
  ).toMatchSnapshot();
});

it('encode', async () => {
  const obj = MsgClaim.fromPartial({
    airdropId: Long.fromInt(1),
    recipient: "cosmos1y54exmx84cqtasvjnskf9fpx9c0rk9tk8fv4",
    conditionType: ConditionType.CONDITION_TYPE_VOTE
  });
  const encoded = MsgClaim.encode(obj).finish();
  const decoded = MsgClaim.decode(encoded);
  expect(encoded instanceof Uint8Array).toBe(true);
  expect(decoded).toMatchSnapshot();
});
