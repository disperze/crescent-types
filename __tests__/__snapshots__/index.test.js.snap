// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`encode 1`] = `
Object {
  "airdropId": Long {
    "high": 0,
    "low": 1,
    "unsigned": true,
  },
  "conditionType": 4,
  "recipient": "cosmos1y54exmx84cqtasvjnskf9fpx9c0rk9tk8fv4",
}
`;

exports[`fromPartial 1`] = `
Object {
  "airdropId": Long {
    "high": 0,
    "low": 1,
    "unsigned": false,
  },
  "conditionType": 4,
  "recipient": "cosmos1y54exmx84cqtasvjnskf9fpx9c0rk9tk8fv4",
}
`;
